src/tests/tracing/eventpipe/userevents/userevents.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Etlx;
+
+namespace Tracing.Tests.UserEvents
+{
+    public class UserEventsTest
+    {
+        private static readonly string trace = "trace.nettrace";
+        private const int SIGINT = 2;
+
+        [DllImport("libc", SetLastError = true)]
+        private static extern int kill(int pid, int sig);
+
+        public static int Main(string[] args)
+        {
+            if (args.Length > 0 && args[0] == "tracee")
+            {
+                UserEventsTracee.Run();
+                return 0;
+            }
+
+            return TestEntryPoint();
+        }
+
+        public static int TestEntryPoint()
+        {
+            string appBaseDir = AppContext.BaseDirectory;
+            string recordTracePath = Path.Combine(appBaseDir, "record-trace");
+            string scriptFilePath = Path.Combine(appBaseDir, "dotnet-common.script");
+            string traceFilePath = Path.Combine(appBaseDir, trace);
+
+            if (!File.Exists(recordTracePath) || !File.Exists(scriptFilePath))
+            {
+                Console.WriteLine("record-trace or dotnet-common.script not found. Test cannot run.");
+                return -1;
+            }
+
+            ProcessStartInfo traceeStartInfo = new();
+            traceeStartInfo.FileName = Process.GetCurrentProcess().MainModule.FileName;
+            traceeStartInfo.Arguments = $"{typeof(UserEventsTest).Assembly.Location} tracee";
+            traceeStartInfo.WorkingDirectory = appBaseDir;
+
+            ProcessStartInfo recordTraceStartInfo = new();
+            recordTraceStartInfo.FileName = recordTracePath;
+            recordTraceStartInfo.Arguments = $"--script-file {scriptFilePath}";
+            recordTraceStartInfo.WorkingDirectory = appBaseDir;
+            recordTraceStartInfo.RedirectStandardOutput = true;
+            recordTraceStartInfo.RedirectStandardError = true;
+
+            using Process traceeProcess = Process.Start(traceeStartInfo);
+            using Process recordTraceProcess = Process.Start(recordTraceStartInfo);
+            recordTraceProcess.OutputDataReceived += (_, args) => Console.WriteLine($"[record-trace] {args.Data}");
+            recordTraceProcess.BeginOutputReadLine();
+            recordTraceProcess.ErrorDataReceived += (_, args) => Console.Error.WriteLine($"[record-trace] {args.Data}");
+            recordTraceProcess.BeginErrorReadLine();
+
+            if (!traceeProcess.HasExited && !traceeProcess.WaitForExit(15000))
```

Comment:
I'd suggest logging that the test is waiting for the tracee to exit, and certainly log if the wait times out and we kill it.